sql/plan/json_table.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/dolthub/vitess/go/vt/sqlparser"
+	"github.com/oliveagle/jsonpath"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+type jsonTablePartition struct {
+	key []byte
+}
+
+var _ sql.Partition = &jsonTablePartition{}
+
+func (j *jsonTablePartition) Key() []byte {
+	return j.key
+}
+
+type jsonTablePartitionIter struct {
+	keys [][]byte
+	pos  int
+}
+
+var _ sql.PartitionIter = &jsonTablePartitionIter{}
+
+func (j *jsonTablePartitionIter) Close(ctx *sql.Context) error {
+	return nil
+}
+
+func (j *jsonTablePartitionIter) Next(ctx *sql.Context) (sql.Partition, error) {
+	if j.pos >= len(j.keys) {
+		return nil, io.EOF
+	}
+
+	key := j.keys[j.pos]
+	j.pos++
+	return &jsonTablePartition{key}, nil
+}
+
+type jsonTableRowIter struct {
+	rows []sql.Row
+	pos  int
+}
+
+var _ sql.RowIter = &jsonTableRowIter{}
+
+func (j *jsonTableRowIter) Next(ctx *sql.Context) (sql.Row, error) {
+	if j.pos >= len(j.rows) {
+		return nil, io.EOF
+	}
+	row := j.rows[j.pos]
+	j.pos++
+	return row, nil
+}
+
+func (j *jsonTableRowIter) Close(ctx *sql.Context) error {
+	return nil
+}
+
+type JSONTable struct {
+	name   string
+	schema sql.PrimaryKeySchema
+	data   []sql.Row
+	rowIdx uint64
+}
+
+var _ sql.Table = &JSONTable{}
```

Comment:
You might consider modelling this as a sql.TableFunction instead

Reply:
I did originally, but I had trouble implementing the `NewInstance` method